File: extra/stress_acia.py
#!/usr/bin/env python3
"""
ACIA Flow Control Stress Test — Host Side

Receives a framed packet from the 6809 on USB CDC port 1 (auxiliary
ACIA), validates it, and echoes the entire frame back only if valid.
Corrupt or incomplete frames are discarded, preventing deadlocks.

Frame format (big-endian):
    SOF(0x02) | SEQ | LEN_H | LEN_L | PAYLOAD[LEN] | CHK | EOF(0x03)
    CHK = XOR of SEQ, LEN_H, LEN_L, and all payload bytes

Protocol:
    1. Host sends 'G' (go)
    2. 6809 replies 'R' (ready) — host reads until it sees 'R'
    3. 6809 sends framed packet; host validates and echoes if good

Usage:
    python3 extra/stress_acia.py /dev/tty.usbmodem*2
"""

import sys
import time
import serial

SOF = 0x02
EOF = 0x03
PAYLOAD_LEN = 4096
NUM_FRAMES = 16
TIMEOUT_S = 30


class BufferedPort:
    """Thin wrapper around a serial.Serial that keeps a pending-byte
    buffer, so leftover bytes from one read phase carry into the next."""

    def __init__(self, port):
        self._port = port
        self._buf = bytearray()

    def read_byte(self, deadline):
        """Return one byte, or None on timeout."""
        while True:
            if self._buf:
                b = self._buf[0]
                del self._buf[0]
                return b
            if time.time() >= deadline:
                return None
            raw = self._port.read(64)
            if raw:
                self._buf.extend(raw)

    def read_n(self, n, deadline):
        """Return exactly n bytes as a bytearray, or None on timeout."""
        while len(self._buf) < n:
            if time.time() >= deadline:
                return None
            raw = self._port.read(min(n - len(self._buf), 4096))
            if raw:
                self._buf.extend(raw)
        result = bytearray(self._buf[:n])
        del self._buf[:n]
        return result

    def push_back(self, data):
        """Push bytes back to the front of the buffer."""
        self._buf[0:0] = data


def recv_frame(bp, deadline, debug=False):
    """Receive and validate one frame from a BufferedPort.
    Returns (seq, payload) on success, or (None, reason_string)."""

    # Wait for SOF — discard any junk before it
    skipped = 0
    while True:
        b = bp.read_byte(deadline)
        if b is None:
            return None, "timeout waiting for SOF"
        if b == SOF:
            break
        if debug:
            print(f"  skip 0x{b:02X}", end="")
        skipped += 1
    if skipped and debug:
        print(f"  ({skipped} bytes skipped before SOF)")

    # SEQ
    b = bp.read_byte(deadline)
    if b is None:
        return None, "timeout reading SEQ"
    seq = b
    chk = seq

    # LEN (big-endian)
    b = bp.read_byte(deadline)
    if b is None:
        return None, "timeout reading LEN_H"
    chk ^= b
    length = b << 8
    b = bp.read_byte(deadline)
    if b is None:
        return None, "timeout reading LEN_L"
    chk ^= b
    length |= b

    if debug:
        print(f"  hdr: seq={seq} len={length}")

    if length > 65535:
        return None, f"length {length} too large"

    # PAYLOAD
    payload = bp.read_n(length, deadline)
    if payload is None:
        return None, f"timeout reading {length}-byte payload"
    for b in payload:
        chk ^= b

    if debug:
        head = ' '.join(f'{b:02X}' for b in payload[:16])
        tail = ' '.join(f'{b:02X}' for b in payload[-4:])
        print(f"  payload[0:16]: {head}")
        print(f"  payload[-4:]:  {tail}")

    # CHK
    b = bp.read_byte(deadline)
    if b is None:
        return None, "timeout reading CHK"
    if b != chk:
        if debug:
            print(f"  chk computed=0x{chk:02X} received=0x{b:02X}")
        return None, f"checksum mismatch: expected 0x{chk:02X}, got 0x{b:02X}"

    # EOF
    b = bp.read_byte(deadline)
    if b is None:
        return None, "timeout reading EOF"
    if b != EOF:
        return None, f"bad EOF marker: 0x{b:02X}"

    return seq, payload


def send_frame(port, seq, payload):
    """Send a framed packet."""
    length = len(payload)
    chk = seq ^ (length >> 8) ^ (length & 0xFF)
    for b in payload:
        chk ^= b

    frame = bytearray()
    frame.append(SOF)
    frame.append(seq)
    frame.append(length >> 8)
    frame.append(length & 0xFF)
    frame.extend(payload)
    frame.append(chk)
    frame.append(EOF)

    port.write(frame)
    port.flush()


def verify_payload(payload, expected_start):
    """Check that payload is the expected incrementing byte pattern.
    Returns (ok, error_count, first_error_str).
    expected_start is the first expected byte value (wraps at 256)."""
    errors = 0
    first_err = ""
    expected = expected_start
    for i, b in enumerate(payload):
        if b != (expected & 0xFF):
            if errors == 0:
                first_err = (f"@{i}: exp=0x{expected & 0xFF:02X} "
                             f"got=0x{b:02X}")
            errors += 1
            expected = b  # resync
        expected += 1
    return errors == 0, errors, first_err


def main():
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <serial-port>")
        print(f"  e.g. {sys.argv[0]} /dev/tty.usbmodem12102")
        sys.exit(1)

    port_name = sys.argv[1]
    print(f"Opening {port_name} ...")
    port = serial.Serial(port_name, timeout=0.05, write_timeout=30)
    port.dtr = True
    time.sleep(0.1)

    # Drain any stale data from previous runs
    port.reset_input_buffer()

    bp = BufferedPort(port)

    # Send 'G' (go) — 6809 replies with 'R' (ready)
    print("Sending start signal ('G') ...")
    port.write(b'G')
    port.flush()

    # Wait for 'R' — proves 6809 received 'G' and is about to send.
    # Any bytes after 'R' stay in the BufferedPort for recv_frame.
    sync_deadline = time.time() + 10.0
    synced = False
    while time.time() < sync_deadline:
        b = bp.read_byte(sync_deadline)
        if b is None:
            break
        if b == ord('R'):
            synced = True
            break
    if not synced:
        print("ERROR: never received 'R' from 6809")
        port.close()
        sys.exit(1)
    print("Synchronised with 6809")

    print(f"Expecting {NUM_FRAMES} x {PAYLOAD_LEN}-byte frames\n")

    start = time.time()
    deadline = start + TIMEOUT_S

    frames_ok = 0
    frames_fail = 0
    total_bytes = 0
    total_errors = 0
    expected_byte = 0  # tracks incrementing pattern across frames

    for expected_seq in range(NUM_FRAMES):
        seq, result = recv_frame(bp, deadline)

        if seq is None:
            print(f"[{expected_seq}] RX failed: {result}")
            frames_fail += 1
            break

        payload = result
        print(f"[{seq}] RX {len(payload)}B ", end="", flush=True)

        # Validate payload
        ok, errors, first_err = verify_payload(payload, expected_byte)
        expected_byte = (expected_byte + len(payload)) & 0xFF
        total_errors += errors
        total_bytes += len(payload)

        if not ok:
            print(f"ERR({errors}: {first_err}) ", end="")

        if seq != expected_seq:
            print(f"SEQ err(exp={expected_seq}) ", end="")

        # Echo the validated frame back
        send_frame(port, seq, payload)
        print("OK")
        frames_ok += 1

        # Extend deadline after each successful round-trip
        deadline = time.time() + TIMEOUT_S

    elapsed = time.time() - start
    port.close()

    # Summary
    print()
    print("--- Host Results ---")
    print(f"Frames OK:      {frames_ok} / {NUM_FRAMES}")
    print(f"Frames failed:  {frames_fail}")
    print(f"Total bytes:    {total_bytes}")
    print(f"Payload errors: {total_errors}")
    print(f"Elapsed:        {elapsed:.2f}s")
    if total_bytes > 0 and elapsed > 0:
        print(f"Throughput:     {total_bytes / elapsed:.0f} bytes/s")
    print()
    if frames_ok == NUM_FRAMES and total_errors == 0:
        print("PASS")
    else:
        print("*** FAIL ***")


if __name__ == "__main__":
    main()

File: chunk/stress_acia.c
/*
 * Copyright (c) 2025-2026 Mark R V Murray
 *
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * ACIA Flow Control Stress Test (6809 side)
 *
 * Runs on the AUXILIARY ACIA ($FFC5/$FFC6, USB CDC port 1).
 * Console ACIA ($FFC3/$FFC4, USB CDC port 0) is used for status output.
 *
 * Protocol:
 *   - Waits for 'G' (go) from host on aux ACIA (30s timeout)
 *   - Sends 'R' (ready) back so host can synchronise
 *   - TX: sends a framed packet (SOF + LEN + payload + CHK + EOF)
 *   - RX: expects the host to echo the same frame back
 *   - If the echo frame is invalid, the whole frame is discarded
 *   - Uses OUTCH/INCH on the aux ACIA (works with IRQ or polled I/O)
 *   - Reports progress and results on the console
 *
 * Frame format (big-endian):
 *   SOF(0x02) | SEQ | LEN_H | LEN_L | PAYLOAD[LEN] | CHK | EOF(0x03)
 *   CHK = XOR of SEQ, LEN_H, LEN_L, and all payload bytes
 *
 * The host-side Python script (extra/stress_acia.py) buffers and
 * validates each frame before echoing, so partial/corrupt data
 * is discarded rather than echoed.
 */

#include <cmoc.h>
#include "pico.h"

/* Frame markers */
#define SOF     0x02
#define FRAME_EOF 0x03

/* Test parameters */
#define PAYLOAD_LEN 4096u
#define NUM_FRAMES  16u
#define TIMEOUT     3000000UL	/* ~30s at 1 MHz E clock */

/* Buffer for block receive — verified in C after the tight asm loop. */
static uint8_t rx_payload[PAYLOAD_LEN];

/* --- I/O helpers (work with both IRQ and polled io.asm) ---------- */

/* Send one byte via OUTCH (blocks until sent/queued).
 * Returns 0 always. */
static uint8_t
aux_send(uint8_t b)
{
	asm {
		EXTERN OUTCH, AUX_PORT
		lda :b
		pshs u
		ldu #AUX_PORT
		lbsr OUTCH
		puls u
	}
	return 0;
}

/* Receive one byte via INCH (polls with timeout).
 * Returns 0 on success, result stored through *out. */
static uint8_t
aux_recv(uint8_t *out)
{
	uint8_t val;
	uint32_t t = TIMEOUT;
	while (t > 0) {
		uint8_t got = 0;
		asm {
			EXTERN INCH, AUX_PORT
			pshs u
			ldu #AUX_PORT
			lbsr INCH
			puls u
			bcc @no
			sta :val
			inc :got
@no
		}
		if (got) {
			*out = val;
			return 0;
		}
		t--;
	}
	return 1;
}

/* --- Frame TX / RX ----------------------------------------------- */

/* Send one framed packet: SOF SEQ LEN_H LEN_L PAYLOAD CHK EOF.
 * Payload is the incrementing byte pattern starting at *tx_val.
 * CHK covers SEQ + LEN + payload.
 * Returns 0 on success (timeout otherwise). */
static uint8_t
send_frame(uint8_t seq, uint16_t len, uint8_t *tx_val)
{
	uint8_t chk = 0;
	uint8_t b;

	if (aux_send(SOF))       return 1;
	chk ^= seq;
	if (aux_send(seq))       return 1;
	b = len >> 8;  chk ^= b;
	if (aux_send(b))         return 1;
	b = len & 0xFF; chk ^= b;
	if (aux_send(b))         return 1;

	/* Payload — tight asm block, ~47 cycles/byte vs ~120 in C.
	 * OUTCH preserves B,X,Y,U.  A may be destroyed on the O@D
	 * drain path, so we save/restore it around each call.
	 * 6809 has no register-to-register XOR, so we XOR via the stack. */
	{
		uint8_t sv = *tx_val;
		uint8_t payload_chk = 0;
		uint16_t count = len;
		asm {
			EXTERN OUTCH, AUX_PORT
			lda :sv
			ldx :count
			clrb
			pshs u
			ldu #AUX_PORT
@sloop			pshs a
			eorb ,s
			lbsr OUTCH
			puls a
			inca
			leax -1,x
			bne @sloop
			puls u
			sta :sv
			stb :payload_chk
		}
		*tx_val = sv;
		chk ^= payload_chk;
	}

	if (aux_send(chk))       return 1;
	if (aux_send(FRAME_EOF)) return 1;
	return 0;
}

/* Receive one framed echo packet.  Verifies SOF, sequence number,
 * length, payload pattern, checksum, and EOF.
 * Returns 0 on full pass; sets *errors, *got on any mismatch. */
static uint8_t
recv_frame(uint8_t expected_seq, uint16_t expected_len,
	   uint8_t *rx_expected, uint16_t *got, uint16_t *errors)
{
	uint8_t b, chk = 0;
	uint16_t len, i;

	/* SOF */
	if (aux_recv(&b))  return 1;
	if (b != SOF) {
		(*errors)++;
		return 2;
	}

	/* SEQ */
	if (aux_recv(&b))  return 1;
	chk ^= b;
	if (b != expected_seq) {
		if (*errors < 10)
			printf("SEQ err: exp=%02X got=%02X\n",
				(unsigned)expected_seq, (unsigned)b);
		(*errors)++;
	}

	/* LEN (big-endian) */
	if (aux_recv(&b))  return 1;
	chk ^= b;
	len = (uint16_t)b << 8;
	if (aux_recv(&b))  return 1;
	chk ^= b;
	len |= b;
	if (len != expected_len) {
		(*errors)++;
		return 3;
	}

	/* PAYLOAD — tight asm receive into rx_payload[], then verify in C.
	 * INCH preserves B,X,Y,U; returns byte in A, carry set = got byte.
	 * No per-byte timeout: spins until each byte arrives.  The Python
	 * host script has its own 30s timeout if something goes wrong.
	 * Checksum is computed in the C verification loop to keep the
	 * asm receive loop as tight as possible (~36 cycles/byte). */
	{
		uint16_t count = len;
		uint8_t *buf = rx_payload;
		asm {
			EXTERN INCH, AUX_PORT
			ldx :buf
			ldy :count
			pshs u
			ldu #AUX_PORT
@rloop			lbsr INCH
			bcc @rloop
			sta ,x+
			leay -1,y
			bne @rloop
			puls u
		}
	}
	/* Verify pattern and accumulate checksum (in-memory, no I/O). */
	{
		uint8_t exp_val = *rx_expected;
		uint16_t local_got = *got;
		uint16_t local_errors = *errors;
		for (i = 0; i < len; i++) {
			chk ^= rx_payload[i];
			if (rx_payload[i] != exp_val) {
				if (local_errors < 10)
					printf("ERR @%u: exp=%02X got=%02X\n",
						local_got, (unsigned)exp_val,
						(unsigned)rx_payload[i]);
				local_errors++;
				exp_val = rx_payload[i];
			}
			exp_val++;
			local_got++;
		}
		*rx_expected = exp_val;
		*got = local_got;
		*errors = local_errors;
	}

	/* CHK */
	if (aux_recv(&b))  return 1;
	if (b != chk) {
		(*errors)++;
		return 4;
	}

	/* EOF */
	if (aux_recv(&b))  return 1;
	if (b != FRAME_EOF) {
		(*errors)++;
		return 5;
	}

	return 0;
}

/* ----------------------------------------------------------------- */

int
main(void)
{
	uint8_t tx_val, rx_expected, ch, seq;
	uint16_t rx_got, rx_errors;
	uint16_t frame_ok, frame_fail;
	uint8_t send_rc, recv_rc;

	/* Flush stale data from aux ACIA BEFORE enabling IRQs.
	 * Repeatedly reset the ACIA (clears the 256-byte SpscQueue on
	 * the Pico side), drain the data register, and wait for
	 * uart_port_pump() to refill from USB.  Loop until no new data
	 * arrives, meaning USB buffers are exhausted.
	 * IRQs are still disabled here (startup default), so no ISR runs. */
	asm {
		pshs a,x
@flush		lda #$03
		sta $FFC5	master reset: clears SpscQueue
		ldx #2000	wait ~6ms for uart_port_pump to refill
@wait		leax -1,x
		bne @wait
@drn		lda $FFC5	drain staged byte
		bita #$01	RDRF?
		beq @chk
		lda $FFC6	discard
		bra @drn
@chk		ldx #500	wait ~1.5ms to see if more data arrives
@chk2		lda $FFC5
		bita #$01
		bne @flush	new data! reset and drain again
		leax -1,x
		bne @chk2
		puls a,x	no more stale data
	}

	/* Now initialise both ACIAs (console + aux).
	 * INIT_AUX does its own C_RESET then enables RX IRQ. */
	initialiseFastSerial();
	setConsoleOutHook(newOutputRoutine);

	printf("\nACIA Flow Control Stress Test\n");
	printf("============================\n");
	printf("Aux ACIA ($FFC5/$FFC6), %u x %u byte frames\n\n",
		NUM_FRAMES, PAYLOAD_LEN);

	/* Wait for 'G' (go) from host script, ~30s timeout. */
	printf("Waiting for host (aux ACIA)...\n");
	{
		uint32_t timeout = TIMEOUT;
		while (timeout > 0) {
			uint8_t have = 0;
			asm {
				EXTERN INCH, AUX_PORT
				pshs u
				ldu #AUX_PORT
				lbsr INCH
				puls u
				bcc @nope
				sta :ch
				inc :have
@nope
			}
			if (have && ch == 'G')
				break;
			timeout--;
		}
		if (timeout == 0) {
			printf("TIMEOUT waiting for host\n");
			SYNC;
			return 1;
		}
	}

	/* Drain any trailing stale data after finding 'G'.
	 * The host won't send anything more until it sees 'R',
	 * so anything arriving now is definitely stale. */
	{
		uint16_t quiet = 0;
		while (quiet < 10000) {
			uint8_t have = 0;
			asm {
				EXTERN INCH, AUX_PORT
				pshs u
				ldu #AUX_PORT
				lbsr INCH
				puls u
				bcc @stl
				inc :have
@stl
			}
			if (have)
				quiet = 0;
			else
				quiet++;
		}
	}

	/* Send 'R' (ready) back to host so it can synchronise. */
	aux_send('R');

	printf("Host connected\n");

	/* --- Send frames, receive echo for each --- */
	tx_val = 0;
	rx_expected = 0;
	rx_got = 0;
	rx_errors = 0;
	frame_ok = 0;
	frame_fail = 0;

	for (seq = 0; seq < NUM_FRAMES; seq++) {
		printf("[%u] ", (unsigned)seq);

		send_rc = send_frame(seq, PAYLOAD_LEN, &tx_val);
		if (send_rc) {
			printf("TX timeout\n");
			frame_fail++;
			break;
		}

		recv_rc = recv_frame(seq, PAYLOAD_LEN, &rx_expected,
				     &rx_got, &rx_errors);
		if (recv_rc) {
			printf("RX err rc=%u\n", (unsigned)recv_rc);
			frame_fail++;
			break;
		}

		printf("OK\n");
		frame_ok++;
	}

	/* Final report */
	printf("\n--- Results ---\n");
	printf("Frames OK:  %u / %u\n", frame_ok, NUM_FRAMES);
	printf("Frames bad: %u\n", frame_fail);
	printf("RX total:   %u bytes\n", rx_got);
	printf("RX errors:  %u\n", rx_errors);

	if (frame_ok == NUM_FRAMES && rx_errors == 0)
		printf("\nPASS\n");
	else
		printf("\n*** FAIL ***\n");

	/* Halt */
	SYNC;
	return 0;
}
